CONTRIBUTING.adoc

```diff
@@ -127,14 +129,36 @@ You need to do a few things when adding a new package, in the following order:
 . Add the package to link:cabal.project[`cabal.project`].
 . Check that you can run `nix build -f default.nix plutus.haskell.projectPackages.<package name>` successfully.
 
-[[update-haskell-pins]]
-=== How to update our pinned Haskell dependencies
+[[update-haskell-deps]]
+=== How to update our Haskell dependencies
 
-We have pinned versions of some Haskell packages specified via the usual `source-repository-package` (Cabal) mechanism.
+Our Haskell packages come from two package repositories:
+- Hackage
+- https://github.com/input-output-hk/cardano-haskell-packages[CHaP] (which is essentially another Hackage)
 
-These can be managed normally, but ensure that:
+The "index state" of each repository is pinned to a particular time in `cabal.project`.
+This tells Cabal to treat the repository "as if" it was the specified time, ensuring reproducibility.
+If you want to use a package version from repository X which was added after the pinned index state time, you need to bump the index state for X.
+This is not a big deal, since all it does is change what packages `cabal` considers to be available when doing solving, but it will change what package versions cabal picks for the plan, and so will likely result in significant recompilation, and potentially some breakage.
+That typically just means that we need to fix the breakage (and add a lower-bound on the problematic package), or add an upper-bound on the problematic package.
 
-* If it is an `source-repository-package`/`extra-dep` from Git, you update the `sha256` mapping in `nix/pkgs/haskell/haskell.nix`.
+Note that `cabal` itself keeps track of what index states it knows about, so when you bump the pinned index state you may need call `cabal update` in order for `cabal` to be happy.
+
+The Nix code which builds our packages also cares about the index state.
+This is represented by some pinned inputs in our flake (see xref:update-nix-pins[here] for more details)
+You can update these by running:
+- `nix flake lock --update-input hackage-nix` for Hackage
+- `nix flake lock --update-input CHaP` for CHaP
+
+==== Use of `source-repository-package`s
+
+We *can* use Cabal's `source-repository-package` mechanism to pull in un-released package versions.
+However, we should try and avoid this.
+In particular, we should not release our packages while we depend on a `source-repository-package`.
+
+If we are stuck in a situation where we need a long-running fork of a package, we should release it to CHaP instead (see the https://github.com/input-output-hk/cardano-haskell-packages[CHaP README] for more).
+
+If you do add a `source-repository-package`, you need to update the `sha256` mapping in `nix/pkgs/haskell/haskell.nix`.
```

Comment:
`plutus-project.nix` also has those shas; does it also need to be updated?

Reply:
Yes, I'll add a TODO for myself to update this section of the doc once the old nix setup goes away.